builder/dockerfile/internals.go

```diff
@@ -0,0 +1,688 @@
+
+	// Work in daemon-specific OS filepath semantics
+	origPath = filepath.FromSlash(origPath)
+	destPath = filepath.FromSlash(destPath)
+
+	if origPath != "" && origPath[0] == os.PathSeparator && len(origPath) > 1 {
+		origPath = origPath[1:]
+	}
+	origPath = strings.TrimPrefix(origPath, "."+string(os.PathSeparator))
+
+	// Twiddle the destPath when its a relative path - meaning, make it
+	// relative to the WORKINGDIR
+	if !system.IsAbs(destPath) {
+		hasSlash := strings.HasSuffix(destPath, string(os.PathSeparator))
+		destPath = filepath.Join(string(os.PathSeparator), filepath.FromSlash(b.runConfig.WorkingDir), destPath)
+
+		// Make sure we preserve any trailing slash
+		if hasSlash {
+			destPath += string(os.PathSeparator)
+		}
+	}
+
+	// Deal with wildcards
+	if allowWildcards && containsWildcards(origPath) {
+		var copyInfos []copyInfo
+		if err := b.context.Walk(func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.Name() == "" {
+				return nil
+			}
+			if match, _ := filepath.Match(origPath, path); !match {
+				return nil
+			}
+
+			// Note we set allowWildcards to false in case the name has
+			// a * in it
+			subInfos, err := b.calcCopyInfo(cmdName, path, destPath, allowLocalDecompression, false)
+			if err != nil {
+				return err
+			}
+			copyInfos = append(copyInfos, subInfos...)
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+		return copyInfos, nil
+	}
+
+	// Must be a dir or a file
+
+	// TODO(tibor): investigate how a path could exist in the context
+	// but have GetFile in tarsum return nil. In that case the hash should be set to origPath.
+	fi, err := b.context.Stat(origPath)
+	if err != nil {
+		return nil, err
+	}
+	copyInfos := []copyInfo{{FileInfo: fi, decompress: allowLocalDecompression}}
+
+	hfi, handleHash := fi.(builder.Hashed)
+	if !handleHash {
+		return copyInfos, nil
+	}
+
+	// Deal with the single file case
+	if !fi.IsDir() {
+		hfi.SetHash("file:" + hfi.Hash())
+		return copyInfos, nil
+	}
+
+	// Must be a dir
+	var subfiles []string
+
+	pfi, ok := fi.(builder.Pather)
+	if !ok {
+		// TODO: can we relax this requirement somehow?
+		return nil, fmt.Errorf("builder: file info (type %T) does not implement builder.Pather", fi)
+	}
+	absOrigPath := pfi.Path()
+
+	// Add a trailing / to make sure we only pick up nested files under
+	// the dir and not sibling files of the dir that just happen to
+	// start with the same chars
+	if !strings.HasSuffix(absOrigPath, string(os.PathSeparator)) {
+		absOrigPath += string(os.PathSeparator)
+	}
+
+	// Need path w/o slash too to find matching dir w/o trailing slash
+	absOrigPathNoSlash := absOrigPath[:len(absOrigPath)-1]
+
+	b.context.Walk(func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		fi, ok := info.(interface {
+			builder.Pather
+			builder.Hashed
+		})
+		if !ok {
+			return fmt.Errorf("builder: unexpected type (%T) for FileInfo", info)
```

Comment:
This is a case where I would change the type signature of the `WalkFunc` to require a `Hashed`. I would not require `Pather` here, since we get it as a method argument.